### PR TITLE
[issue-278] POC implementing protocol buffer message format.

### PR DESCRIPTION
**Pull request by Crim**
### Note
This is an **extremely rough proof of concept** code hacked together the quickest way possible for #278.  Not intended to be used/merged as is without considerable work/cleanup.

KLIP-0 #1472 

### Description
Adds the ability to run ksql over messages encoded with protocol buffers.

### Testing done 
Only simple unit tests exist now, additional test cases to be written.

### Reviewer checklist
- [X] Replace JSON pass-thru hack with proper implementation
- [ ] Add type checking to protobuf conversion class
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-serde/src/main/java/io/confluent/ksql/serde/protobuf/KsqlProtobufSerializer.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      throw new RuntimeException(e.getMessage(), e);
+    }
+
+    ksqlJsonSerializer.configure(props, isKey);
+  }
+
+  @Override
+  public byte[] serialize(final String topic, final GenericRow data) {
+    // Convert from GenericRow into JSON (bytes)
+    final byte[] jsonBytes = ksqlJsonSerializer.serialize(topic, data);
```
Comment (Crim) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/protobuf/KsqlProtobufSerializer.java]: Quick and dirty hack just to see this work w/ minimal code.  Convert the GenericRow into JSON, and use an off-the-shelf library to hydrate json back to the protobuf.

Ideally we would just translate directly to protobuf.

File: ksql-serde/src/main/java/io/confluent/ksql/serde/protobuf/KsqlProtobufDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    try {
+      // Parse Protobuf from the bytes
+      final MessageOrBuilder message
+        = (MessageOrBuilder) parseFromMethod.invoke(protobufType, bytes);
+
+      // Convert to json
+      final String jsonStr = jsonFormatPrinter.print(message);
+
+      // Run JSON thru JSON Deserializer
```
Comment (Crim) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/protobuf/KsqlProtobufDeserializer.java]: Quick and dirty hack just to see this work w/ minimal code. Convert the protobuf into JSON using an off-the-shelf library, and then push that through the JSON Deserializer to get a GenericRow.

Ideally we would just do the translation directly.

**Comment by apurvam** [comment]
Thanks @Crim ! As mentioned on the issue, it would be great to submit a KLIP instead of writing a google doc. The process is documented [here](https://github.com/confluentinc/ksql/tree/master/design-proposals). You will be immortalized as being the author of KLIP-0 :)

**Comment by arbiter34** [comment]
I fleshed out protobuf support based on @Crim's branch: https://github.com/Crim/ksql/pull/4

I believe this covers every case I'm aware of, we've been running it in our test environment successfully including CSAS support with (value_format='PROTOBUF', etc).

I'm not sure on the status of this feature but it'd be nice to see it hit mainline.

**Comment by MichaelDrogalis** [comment]
Given that this has languished and still has large design issues, I'd like to close this. We're more than happy to try to get a patch for this in again, but it needs some rethinking.

**Pull request closed without merging**